pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -831,17 +839,13 @@ type nominator struct {
 	sync.RWMutex
 }
 
-func (npm *nominator) add(pi *framework.PodInfo, nodeName string) {
-	// always delete the pod if it already exist, to ensure we never store more than
+func (npm *nominator) add(pi *framework.PodInfo, nodeName *string) {
+	// Always delete the pod if it already exists, to ensure we never store more than
 	// one instance of the pod.
 	npm.delete(pi.Pod)
 
-	nnn := nodeName
-	if len(nnn) == 0 {
-		nnn = NominatedNodeName(pi.Pod)
-		if len(nnn) == 0 {
-			return
-		}
+	if nodeName == nil {
```

Comment:
so `nil` removes the nominated node name?
Wasn't that supposed to be empty string?

Reply:
In the before, if an empty string is passed in, we always look at and honor the nominatedNodeName if it carries. Now the logic gets tweaked:

- if you want to follow the old logic, on the caller side, use `adaptiveNominatedNodeName()` to always honor nominatedNodeName
- if you want to enforce overriding, pass in a non-nil string pointer (no matter it's empty or a concrete string).

Reply:
I'm not liking `adaptiveNominatedNodeName`, as it could be forgotten. I think it should be handled within this function: if passing nil, leave; otherwise override

Reply:
> if passing nil, leave; otherwise override

Passing the string pointer still needs to be decided at the caller side - whether it's a pod with empty nominatedNodeName, so we interpreted it as nil; or, the caller have baked the string pointer properly (like `recordSchedulingFailure()`)

Reply:
Here is my proposal:
- `nil` means do not change anything. Then, we would query the nominatedNodeName from the pod status.
- `""` means clear
- `"foo"` is override

I think that logic can be handled here.

Reply:
It makes sense to me and can save the efforts on the caller side.

One question on `"" means clear`: to keep the behavior with before, "" and "regular string" are treated the same. If we really want to clear the entry, we can do it in a follow-up.

Reply:
I think distinguishing between nil and empty is asking for trouble. I also didn't like the old code that implicitly treated an empty nodeName as a fallback to the value in the pod. Lots of implicit behavior that one can't figure out without carefully parsing the code.

Not sure if this was the old code, but why not simply have the following:

```
// Call the following in the places were you are currently passing in nil.
add(pi) {
 addToNode(pi, pi.Pod.Status.NominatedNodeName)
}
```

```
// Alway adds the pod to the passed nodeName
addToNode(pi, nodeName) {
 ...
}
```

Reply:
> Not sure if this was the old code, but why not simply have the following

This is basically the previous version except for that some places needs to inject NominatedNodeName adapatively.